gandy/tests/test_models/test_gans.py

```diff
@@ -0,0 +1,200 @@
+        generator and discriminator are attributes with type == Keras model.
+        """
+        # CHECK (normal) GAN
+        # create gan instance
+        subject = gans.GAN(xshape=(4,), yshape=(2,))
+        kwargs = dict(noise_shape=(5,))
+        subject._build(kwargs)
+        # assert create generator function called
+        subject._model.create_generator.assert_called_once_with(kwargs)
+        # assert create discriminator function called
+        subject._model.create_discriminator.assert_called_once_with(kwargs)
+        # check attributes
+        self.assertTrue(subject.conditional, False)
+        self.assertTrue(subject.noise_shape, (5,))
+
+        # CHECK Conditional GAN
+        # create gan instance
+        subject = gans.CondGAN(xshape=(4,), yshape=(2, 4))
+        assert issubclass(gans.CondGAN, gans.GAN)
+        kwargs = dict(noise_shape=(5,), n_classes=4)
+        subject._build(kwargs)
+        # assert create generator function called
+        subject._model.create_generator.assert_called_once_with(kwargs)
+        # assert create discriminator function called
+        subject._model.create_discriminator.assert_called_once_with(kwargs)
+        # check attributes
+        self.assertTrue(subject.conditional, True)
+        self.assertTrue(subject.noise_shape, (5,))
+        self.assertTrue(subject.n_classes, 4)
+        return
+
+    def test__train(self):
+        """
+        Test train function.
+
+        The train function calls the fit function for the generator and the
+        predict function for the discriminator.
+        This checks that there is a Keras callback History object returned.
+        """
+        Xs = 'Xs'
+        Ys = 'Ys'
+        subject = gans.GAN(xshape=(4,), yshape=(2,))
+        subject.iterbacthes = mock.MagicMock(name='iterbatches',
+                                             return_value="Batch1")
+        subject._model.fit_gan = mock.MagicMock(name='fit_gan')
+        kwargs = dict(option='x1')
+        subject._train(Xs, Ys, kwargs)
+
+        # assert fit_gan was called
+        subject.iterbacthes.assert_called_with(Xs, Ys, kwargs)
+        subject._model.fit_gan.assert_called_with("Batch1")
+        return
+
+    @unittest.mock.patch('gandy.models.gans.GAN._build', return_value='Model')
+    def test__predict(self, mocked__build):
+        """
+        Test predict function.
+
+        The predict function returns predictions and uncertainties.
+        This checks predictions and uncertainties are the appropriate shape
+        and the appropriate deepchem calls are made.
+        """
+        Xs = 'Xs'
+        # CHECK (normal) GAN
+        subject = gans.GAN(xshape=(4,), yshape=(2,))
+        subject.predict_gan_generator = mock.MagicMock(
+            name='predict_gan_generator', return_value='generated_points')
+        preds, ucs = subject._predict(Xs)
+        subject._predict.assert_called_with(Xs)
```

Comment:
This is redundant, you see in the line above that it is. Also it is not mocked. You want to mock nested functions, but not the function that is being tested itself.